torch/distributed/pipelining/_backward.py

```diff
@@ -1,12 +1,234 @@
 # mypy: allow-untyped-defs
 # Copyright (c) Meta Platforms, Inc. and affiliates
-from typing import List, Optional
+import collections
+import logging
+import weakref
+from typing import Any, cast, Deque, Dict, Iterator, List, Optional, Set, Tuple, Union
 
 import torch
+from torch.autograd.graph import GradientEdge, Node
+from torch.nn import Parameter
 
 from ._debug import map_debug_info
 
 
+logger = logging.getLogger(__name__)
+
+
+def _get_grad_fn_or_grad_acc(t: torch.Tensor) -> Union[Node, None]:
+    if t.requires_grad and t.grad_fn is None:
+        # if no grad function (leaf tensors) we use view
+        viewed_t = t.view_as(t)
+        grad_fn = viewed_t.grad_fn
+        if grad_fn is not None:
+            return grad_fn.next_functions[0][0]
+        else:
+            raise RuntimeError("grad_fn after adding view is None")
+    else:
+        return t.grad_fn
+
+
+def reverse_closure(
+    roots: List[Node], target_nodes: Set[Node]
+) -> Tuple[Set[Node], Set[Node]]:
+    # Recurse until we reach a target node
+    closure: Set[Node] = set()
+    actual_target_nodes = set()
+    q: Deque[Node] = collections.deque()
+    for node in roots:
+        if node is not None and node not in closure:
+            closure.add(node)
+            q.append(node)
+    while q:
+        node = q.popleft()
+        metadata = cast(Dict[str, List], node.metadata)
+        reverse_edges = metadata.get("reverse_edges", [])
+        for holder_ref, idx in reverse_edges:
+            ref = holder_ref()
+            if ref is None:
+                # this reverse graph is no longer alive
+                # raise RuntimeError("Reverse graph is no longer alive")
+                continue
+            fn = ref.node
+            if fn in closure or fn is None:
+                continue
+            if fn in target_nodes:
+                actual_target_nodes.add(fn)
+                continue
+            closure.add(fn)
+            q.append(fn)
+    return closure, actual_target_nodes
+
+
+# Enable weak pointer
+class Holder:
+    def __init__(self, node: Node):
+        self.node = node
+
+
+# TODO: use weak references to avoid reference cycle
+def construct_reverse_graph(roots: List[Node]) -> List[Holder]:
+    q: Deque[Node] = collections.deque()
+    root_seen: Set[Node] = set()
+    reverse_graph_refs: List[Holder] = []
+    for node in roots:
+        if node is not None and node not in root_seen:
+            q.append(node)
+            root_seen.add(node)
+    while q:
+        node = q.popleft()
+        for fn, idx in node.next_functions:
+            if fn is not None:
+                # Don't necessarily need to store on the graph
```

Comment:
any downsides to storing it on the graph @soulitzer

Reply:
oh, i think i'm starting to see whats going on.
1- we are storing new reverse pointer metadata on the original graph nodes
2- the reverse pointers are 2 stage for a reason: (a) we store weakrefs (to nodes, or to holders probably doesn't matter?) on the autograd nodes, so we dont create refcycles there.  (b) we also create strong refs to the nodes and hold onto those at least until we're done messing around with this graph, later we delete these to clean things up

Is that all correct?  this should be explained in comments

Reply:
it does feel kinda bad to store stuff on the autograd graph.  after our code runs, the graph is mutated.  someone could notice (?) 

if we didn't want to modify the graph, i guess the alternative is to construct a ReverseGraph object, store a dict of id(node) -> node in it, and then manage the lifetime of this ReverseGraph object, making sure to delete it once we are done with it.  would that work?

Reply:
> it does feel kinda bad to store stuff on the autograd graph. after our code runs, the graph is mutated. someone could notice (

agreed, would be better to avoid using the grad_fn.metadata, and a separate ReverseGraph should work yeah

Reply:
In a follow up PR i am working on I just pass a dictionary to save this state as you mentioned!